docs/tutorials/developer-journey/level-2/2.1-storage-persistence.md

```diff
@@ -0,0 +1,189 @@
+
+To use stable memory requires anticipating and indicating which canister data you want to be retained after a canister upgrade. For some canisters, this data might be all of the canister's data, while for others it may be only certain parts or none. By default, stable memory starts empty, and can hold up to 32GB as long as the subnet the canister is running on has the available space. If a canister uses more stable memory than 32GB, the canister will trap and become unrecoverable. 
+
+## Memory types and terms 
+
+There are several terms associated with memory and storage on the IC. To avoid confusion between them, we'll define them here. 
+
+- **Stable memory:** Stable memory refers to the Internet Computer's long-term data storage feature. Stable memory is not language specific, and can be utilized by canisters written in Motoko, Rust, or any other language. Stable memory can hold up to 32GB if the subnet can accommodate it. 
+
+- **Heap storage:** Heap storage refers to the regular Wasm memory data store for each canister. This storage is temporary and is not persisted across canister upgrades. Data stored in heap storage is removed when the canister is upgraded, stopped, or reinstalled. Heap storage is limited to 4GB. 
+
+- **Stable storage:** Stable storage is a Motoko-specific term that refers to the Motoko stable storage feature. Stable storage uses the IC's stable memory to persist data across canister upgrades. 
+
+- **Stable variables:** Stable variables are a Motoko-specific feature that refers to variables defined in Motoko that use the `stable` modifier which indicates that the value of the variable should be persisted across canister upgrades. 
+
+To further understand stable memory and how to use it, let's learn about upgrading canisters. 
+
+## Upgrading canisters
+
+Once a canister has been deployed and is running, there may need to be changes made to the canister's code to fix bugs or introduce new features. To make these changes, the canister must be upgraded. 
+
+Upgrading a canister is a key feature of the IC, since it allows canister smart contracts to persist using Wasm memory that utilizes the IC's stable memory feature. When a canister is upgraded, the existing state of the canister is preserved while there are changes being made to the canister's code. 
+
+### Motoko stable memory workflow
+
+For canisters written in Motoko, stable memory can be utilized through the Motoko stable storage and stable variable features. These features as most notably used in the canister upgrade process. The upgrade process for Motoko canisters is as follows:
+
+- First, the canister must be stopped so it does not accept any new incoming requests. 
+- A `pre_upgrade` hook is executed if one is defined. This hook can be used for functions such as creating a backup of data. 
```

Comment:
Using the `pre_upgrade` is strongly discouraged. We should have a warning here. The reason is that when you upgrade to a new version, you first run the code in the `pre_upgrade` hook of the current wasm, i.e. if there's a bug in this code you're stuck.